pkg/monitor/agent.go

```diff
@@ -29,43 +29,41 @@ import (
 type agentMonitor struct {
 	client  clientset.Interface
 	querier agentquerier.AgentQuerier
+	// agentCRD is the desired state of agent monitoring CRD which agentMonitor expects.
+	agentCRD *v1beta1.AntreaAgentInfo
 }
 
 // NewAgentMonitor creates a new agent monitor.
 func NewAgentMonitor(client clientset.Interface, querier agentquerier.AgentQuerier) *agentMonitor {
-	return &agentMonitor{client: client, querier: querier}
+	return &agentMonitor{client: client, querier: querier, agentCRD: nil}
 }
 
 // Run creates AntreaAgentInfo CRD first after controller is running.
 // Then updates AntreaAgentInfo CRD every 60 seconds.
 func (monitor *agentMonitor) Run(stopCh <-chan struct{}) {
 	klog.Info("Starting Antrea Agent Monitor")
-	agentCRD := monitor.getAgentCRD()
-	var err error = nil
+	monitor.agentCRD = monitor.getAgentCRD()
 
 	// Initialize agent monitoring CRD.
-	if agentCRD == nil {
-		agentCRD, err = monitor.createAgentCRD()
+	if monitor.agentCRD == nil {
+		agentCRD, err := monitor.createAgentCRD()
 		if err != nil {
-			klog.Errorf("Failed to create agent monitoring CRD %+v: %v", agentCRD, err)
+			klog.Errorf("Failed to create agent monitoring CRD %+v: %v", monitor.agentCRD, err)
 			return
 		}
+		// The monitor.agentCRD is not set until agentCRD is successfully created.
+		monitor.agentCRD = agentCRD
 	} else {
-		agentCRD, err = monitor.updateAgentCRD(agentCRD)
+		agentCRD, err := monitor.updateAgentCRD()
 		if err != nil {
-			klog.Errorf("Failed to update agent monitoring CRD %+v: %v", agentCRD, err)
+			klog.Errorf("Failed to update agent monitoring CRD %+v: %v", monitor.agentCRD, err)
 			return
 		}
+		monitor.agentCRD = agentCRD
```

Comment:
To make it more robust, I think createAgentCRD and  updateAgentCRD could be in the loop too, otherwise a temp disconnection here could lead to missing CRD or no CRD update.

Reply:
Thanks for the suggestion, have updated this part. 
As discussed, I also cover the cases of accidentally deleting monitoring CRDs and issues like mismatch resourceVersion caused by temporary network issues in the latest commit.
Please help take a look.